pyface/toolkit.py

```diff
@@ -49,105 +72,82 @@ def provisional_toolkit(toolkit_name):
 
 
 def _init_toolkit():
-    """ Initialise the current toolkit. """
+    """ Initialise a toolkit, if possible. """
 
     def import_toolkit(tk):
-        try:
-            # Try and import the toolkit's pyface backend init module.
-            be = 'pyface.ui.%s.' % tk
-            __import__(be + 'init')
-        except:
-            raise
-        return be
+        plugins = list(pkg_resources.iter_entry_points('pyface.toolkits', tk))
+        if len(plugins) == 0:
+            msg = "no Pyface plugin found for toolkit '{}'"
+            raise RuntimeError(msg.format(tk))
+        elif len(plugins) > 1:
+            msg = ("multiple Pyface plugins found for toolkit '{}': {}")
+            modules = ', '.join(plugin.module_name for plugin in plugins)
+            logger.warning(msg.format(tk, modules))
+
+        while plugins:
+            plugin = plugins.pop(0)
+            try:
+                tk_object = plugin.load()
+                return tk_object
+            except ImportError as exc:
+                logger.exception(exc)
+                msg = "Could not load plugin '{}' from '{}'"
+                logger.warning(msg.format(plugin.name, plugin.module_name))
+        else:
+            # no success
+            raise exc
 
     # Get the toolkit.
     if ETSConfig.toolkit:
-        be = import_toolkit(ETSConfig.toolkit)
-    else:
-        # Toolkits to check for if none is explicitly specified.
-        known_toolkits = ('qt4', 'wx', 'null')
+        return import_toolkit(ETSConfig.toolkit)
 
-        for tk in known_toolkits:
-            try:
-                with provisional_toolkit(tk):
-                    be = import_toolkit(tk)
-                break
-            except ImportError as exc:
-                msg = "Could not import Pyface backend '{0}'"
-                logger.info(msg.format(tk))
-                if logger.getEffectiveLevel() <= logging.INFO:
-                    logger.exception(exc)
-        else:
-            # Try to import the null toolkit but don't set the ETSConfig toolkit
-            try:
-                be = import_toolkit('null')
-                import warnings
-                msg = ("Unable to import the '{0}' backend for pyface; " +
-                       "using the 'null' backend instead.")
-                warnings.warn(msg.format(toolkit_name), RuntimeWarning)
-            except ImportError as exc:
+    # Try known toolkits first.
+    known_toolkits = ('qt4', 'wx')
+    for tk in known_toolkits:
+        try:
+            with provisional_toolkit(tk):
+                return import_toolkit(tk)
+        except RuntimeError as exc:
+            if logger.getEffectiveLevel() <= logging.INFO:
```

Comment:
It is better to have a single logging message. I would suggest something like

```python
exc_info = logger.getEffectiveLevel() <= logging.INFO
msg = "Could not import Pyface backend %r"
level = 'ERROR' if exc_info else 'INFO'
logger.log(level, msg, tk, excinfo=exc_info)
```

(it is also advised to use `%` for log messages)